File: adk-go/src/application/tools/calculators.go
package tools

import (
	"math"
	"strings"
)

type LineCalculation struct {
	NetTotal   float64 `json:"net_total"`
	VatAmount  float64 `json:"vat_amount"`
	GrossTotal float64 `json:"gross_total"`
	Error      string  `json:"error,omitempty"`
}

type TotalsCalculation struct {
	Net   float64 `json:"net"`
	Vat   float64 `json:"vat"`
	Gross float64 `json:"gross"`
}

func roundTo2Decimals(num float64) float64 {
	return math.Round(num*100) / 100
}

func CalculateLine(quantity float64, unitPriceNet float64, vatRate string) LineCalculation {
	rates := map[string]float64{
		"23%": 0.23, "8%": 0.08, "5%": 0.05, "0%": 0.00, "zw.": 0.00,
	}

	cleanRate := strings.ToLower(strings.ReplaceAll(vatRate, " ", ""))
	rate, exists := rates[cleanRate]
	if !exists {
		return LineCalculation{Error: "Nieznana stawka VAT: " + vatRate}
	}

	netTotal := roundTo2Decimals(quantity * unitPriceNet)
	vatAmount := roundTo2Decimals(netTotal * rate)
	grossTotal := roundTo2Decimals(netTotal + vatAmount)

	return LineCalculation{
		NetTotal:   netTotal,
		VatAmount:  vatAmount,
		GrossTotal: grossTotal,
	}
}

func CalculateTotals(lines []map[string]any) TotalsCalculation {
	var net, vat float64
	for _, l := range lines {
		if n, ok := l["net_total"].(float64); ok {
			net += n
		}
		if v, ok := l["vat_amount"].(float64); ok {
			vat += v
		}
	}
	return TotalsCalculation{
		Net:   roundTo2Decimals(net),
		Vat:   roundTo2Decimals(vat),
		Gross: roundTo2Decimals(net + vat),
	}
}

File: adk-typescript/src/application/tools/calculators.ts
interface LineCalculation {
    net_total: number;
    vat_amount: number;
    gross_total: number;
}

export function calculateLine(quantity: number, unitPriceNet: number, vatRate: string): LineCalculation {
    const netTotal = quantity * unitPriceNet;
    let vatMultiplier = 0.0;
    
    if (vatRate !== "zw" && vatRate !== "np") {
        const cleanRate = vatRate.replace('%', '');
        vatMultiplier = parseFloat(cleanRate) / 100.0;
    }
    
    const vatAmount = Math.round((netTotal * vatMultiplier) * 100) / 100;
    const grossTotal = Math.round((netTotal + vatAmount) * 100) / 100;
    
    return {
        net_total: netTotal,
        vat_amount: vatAmount,
        gross_total: grossTotal
    };
}

interface TotalsCalculation {
    total_net: number;
    total_vat: number;
    total_gross: number;
}

export function calculateTotals(lines: any[]): TotalsCalculation {
    let totalNet = 0.0;
    let totalVat = 0.0;
    
    if (lines && Array.isArray(lines)) {
        for (const line of lines) {
            totalNet += line.net_total || 0;
            totalVat += line.vat_amount || 0;
        }
    }
    
    const totalGross = totalNet + totalVat;
    
    return {
        total_net: Math.round(totalNet * 100) / 100,
        total_vat: Math.round(totalVat * 100) / 100,
        total_gross: Math.round(totalGross * 100) / 100
    };
}